File: app/src/main/java/mydebts/android/app/feature/event/EventFragment.java
package mydebts.android.app.feature.event;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.util.Date;
import java.util.Iterator;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import mydebts.android.app.R;
import mydebts.android.app.data.EventsSource;
import mydebts.android.app.data.ParticipantsSource;
import mydebts.android.app.data.PersonsSource;
import mydebts.android.app.data.model.Event;
import mydebts.android.app.data.model.Participant;
import mydebts.android.app.di.SubcomponentBuilderResolver;
import mydebts.android.app.feature.main.MainRouter;
import mydebts.android.app.rx.RxUtil;

public class EventFragment extends Fragment {
    private static final String ARG_EVENT_ID = "ARG_EVENT_ID";

    private ParticipantsAdapter adapter;

    @Inject RxUtil rxUtil;
    @Inject EventsSource eventsSource;
    @Inject PersonsSource personsSource;
    @Inject ParticipantsSource participantsSource;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);

        ((EventSubcomponent.Builder)SubcomponentBuilderResolver.resolve(this))
                .build()
                .inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.activity_event, container, false);

        RecyclerView listParticipants = (RecyclerView)rootView.findViewById(R.id.list_participants);
        listParticipants.setLayoutManager(new LinearLayoutManager(listParticipants.getContext()));

        adapter = new ParticipantsAdapter();
        listParticipants.setAdapter(adapter);

        if (getArguments() != null && getArguments().containsKey(ARG_EVENT_ID)) {
            participantsSource.getByEventId(getArguments().getLong(ARG_EVENT_ID))
                    .compose(rxUtil.singleSchedulersTransformer())
                    .subscribe(adapter::setItems);
        }

        return rootView;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_event, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_set_date:
                return true;
            case R.id.action_save:
                saveEvent(adapter.getParticipants());
                return true;
            case R.id.action_delete:
                deleteEvent();
                return true;
            default:
                return false;
        }
    }

    private void saveEvent(List<Participant> participants) {
        for (Iterator<Participant> iterator = participants.iterator(); iterator.hasNext();) {
            Participant participant = iterator.next();
            if (TextUtils.isEmpty(participant.getPerson().getName())
                    || Math.abs(participant.getDebt()) < 0.001) {
                iterator.remove();
            }
        }

        if (participants.size() == 0) {
            return;
        }

        Date date = new Date();
        date.setTime(System.currentTimeMillis());

        Observable<Event> eventObservable = eventsSource.insert(Event.builder()
                .name(date.toString())
                .date(date)
                .build())
                .toObservable();

        Observable<Participant> participantObservable = Observable.fromIterable(participants);

        Observable.combineLatest(eventObservable, participantObservable,
                    (event, participant) -> Participant.builder(participant)
                        .event(event)
                        .build())
                .flatMap(participant -> personsSource.insert(participant.getPerson())
                        .map(person -> Participant.builder(participant)
                                        .person(person)
                                        .build())
                        .toObservable())
                .flatMap(participant -> participantsSource.insert(participant)
                                            .toObservable())
                .compose(rxUtil.observableSchedulersTransformer())
                .doOnComplete(() -> ((MainRouter)getActivity()).navigateBack())
                .subscribe();
    }

    private void deleteEvent() {
        if (getArguments().containsKey(ARG_EVENT_ID)) {
            eventsSource.delete(Event.builder()
                                    .id(getArguments().getLong(ARG_EVENT_ID))
                                    .build())
                    .flatMap(event -> participantsSource.deleteByEventId(event.getId()))
                    .compose(rxUtil.singleSchedulersTransformer())
                    .subscribe(obj -> ((MainRouter) getActivity()).navigateBack());
        } else {
            ((MainRouter) getActivity()).navigateBack();
        }
    }

    public static EventFragment newInstance(@NonNull Event event) {
        Bundle args = new Bundle();
        args.putLong(ARG_EVENT_ID, event.getId());

        EventFragment fragment = newInstance();
        fragment.setArguments(args);

        return fragment;
    }

    public static EventFragment newInstance() {
        return new EventFragment();
    }
}

File: app/src/main/java/mydebts/android/app/feature/event/EventFragment.kt
package mydebts.android.app.feature.event

import android.app.DatePickerDialog
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker

import javax.inject.Inject

import mydebts.android.app.R
import mydebts.android.app.data.model.Event
import mydebts.android.app.data.model.Participant
import mydebts.android.app.di.SubcomponentBuilderResolver
import mydebts.android.app.feature.main.MainRouter
import mydebts.android.app.feature.participant.ParticipantDialogFragment

class EventFragment : Fragment(), EventScreen, DatePickerDialog.OnDateSetListener, ParticipantDialogFragment.Callback {

    @Inject lateinit var presenter: EventPresenter

    private lateinit var emptyView: View
    private lateinit var participantsRecyclerView: RecyclerView
    private lateinit var adapter: ParticipantsAdapter

    private var deleteMenuItem: MenuItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)

        (SubcomponentBuilderResolver.resolve(this) as EventSubcomponent.Builder)
                .event(arguments?.getParcelable(ARG_EVENT))
                .fragment(this)
                .build()
                .inject(this)

        adapter = ParticipantsAdapter(presenter.getParticipants())
        adapter.setOnParticipantClickListener { presenter.onParticipantClick(it) }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater?.let {
                val rootView = it.inflate(R.layout.activity_event, container, false)

                emptyView = rootView.findViewById(R.id.text_no_participants)

                participantsRecyclerView = rootView.findViewById(R.id.list_participants)
                participantsRecyclerView.layoutManager = LinearLayoutManager(participantsRecyclerView.context)
                participantsRecyclerView.adapter = adapter

                val addParticipantButton: View = rootView.findViewById(R.id.button_add_participant)
                addParticipantButton.setOnClickListener { presenter.onAddNewParticipantClick() }

                rootView
            }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        presenter.onViewCreated()
    }

    override fun onDestroyView() {
        presenter.onDestroyView()
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.menu_event, menu)
        deleteMenuItem = menu?.findItem(R.id.action_delete)

        presenter.onCreateOptionsMenu()
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean =
        item?.let {
            return when (it.itemId) {
                R.id.action_set_date -> {
                    presenter.onSetDateClick()
                    true
                }
                R.id.action_save -> {
                    presenter.onActionSaveClick()
                    true
                }
                R.id.action_delete -> {
                    presenter.onActionDeleteClick()
                    true
                }
                else -> {
                    false
                }
            }
        } ?: false

    override fun onParticipantResult(participant: Participant) {
        presenter.addParticipant(participant)
    }

    override fun showTitle(title: CharSequence) {
        activity.title = title
    }

    override fun setDeleteMenuItemVisible(visible: Boolean) {
        deleteMenuItem?.isVisible = visible
    }

    override fun showDatePicker(year: Int, month: Int, day: Int) {
        val datePickerDialog = DatePickerDialog(activity, 0, this, year, month, day)
        datePickerDialog.show()
    }

    override fun setEmptyViewVisibility(visibility: Int) {
        emptyView.visibility = visibility
    }

    override fun setParticipantsViewVisibility(visibility: Int) {
        participantsRecyclerView.visibility = visibility
    }

    override fun notifyParticipantsChanged() {
        adapter.notifyDataSetChanged()
    }

    override fun notifyParticipantInserted(position: Int) {
        adapter.notifyItemInserted(position)
    }

    override fun showParticipant(participant: Participant) {
        ParticipantDialogFragment.newInstance(participant).show(childFragmentManager, null)
    }

    override fun showNewParticipant() {
        ParticipantDialogFragment.newInstance().show(childFragmentManager, null)
    }

    override fun navigateBack() {
        (activity as MainRouter).navigateBack()
    }

    override fun onDateSet(view: DatePicker?, year: Int, month: Int, dayOfMonth: Int) {
        presenter.setDate(year, month, dayOfMonth)
    }

    companion object {
        private val ARG_EVENT = "ARG_EVENT"

        fun newInstance(event: Event): EventFragment {
            val fragment = newInstance()
            fragment.arguments = Bundle()
            fragment.arguments.putParcelable(ARG_EVENT, event)
            return fragment
        }

        fun newInstance(): EventFragment = EventFragment()
    }
}
